op-node/rollup/driver/state.go

```diff
@@ -408,87 +291,68 @@ func (s *state) loop() {
 
 			// We need to catch up to the next origin as quickly as possible. We can do this by
 			// requesting a new block ASAP instead of waiting for the next tick.
-			// TODO: If we want to consider confirmations, need to consider here too.
-			if s.l1Head.Number > s.l2Head.L1Origin.Number {
+			// We don't request a block if the confirmation depth is not met.
+			if s.l1Head.Number > s.l2Head.L1Origin.Number+s.DriverConfig.SequencerConfDepth {
 				s.log.Trace("Asking for a second L2 block asap", "l2Head", s.l2Head)
 				// But not too quickly to minimize busy-waiting for new blocks
 				time.AfterFunc(time.Millisecond*10, reqL2BlockCreation)
 			}
 
 		case payload := <-s.unsafeL2Payloads:
-			s.log.Info("Optimistically processing unsafe L2 execution payload", "id", payload.ID())
-			err := s.handleUnsafeL2Payload(ctx, payload)
-			if err != nil {
-				s.log.Warn("Failed to process L2 execution payload received from p2p", "err", err)
-			}
+			s.snapshot("New unsafe payload")
+			s.log.Info("Optimistically queueing unsafe L2 execution payload", "id", payload.ID())
+			s.derivation.AddUnsafePayload(payload)
+			reqStep()
 
 		case newL1Head := <-s.l1Heads:
+			s.log.Info("new l1 Head")
 			s.snapshot("New L1 Head")
-			ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
-			err := s.handleNewL1Block(ctx, newL1Head)
-			cancel()
-			if err != nil {
-				s.log.Error("Error in handling new L1 Head", "err", err)
-			}
-
-			// The block number of the L1 origin for the L2 safe head is at least SeqWindowSize
-			// behind the L1 head. We can therefore attempt to shift the safe head forward by at
-			// least one L1 block. If the node is holding on to unsafe blocks, this may trigger a
-			// reorg on L2 in the case that safe (published) data conflicts with local unsafe
-			// block data.
-			if s.l1Head.Number-s.l2SafeHead.L1Origin.Number >= s.Config.SeqWindowSize {
-				s.log.Trace("Requesting next step", "l1Head", s.l1Head, "l2Head", s.l2Head, "l1Origin", s.l2SafeHead.L1Origin)
-				reqStep()
-			}
-
+			s.handleNewL1Block(newL1Head)
```

Comment:
How does this handle reorgs?

Reply:
We don't handle reorgs unless we actually hit a L1 block on traversal. I.e. we currently don't immediately handle reorgs to shorter L1 chains. But the L2 won't grow one step further either without detecting the reorg (i.e. as soon as it can get L1 input N+1 it will see it doesn't build on N and then the pipeline will reset).

Reply:
ok, I've figured out how reorgs are being handled & it makes sense to me